### ISSUE TITLE
Add lazy chaining support to _.reduce

### ISSUE THREAD

**Issue by mehranhatami** [mehranhatami/lodash#1]
@jdalton: I started coding to add lazy chaining support to `_.reduce` yesterday and I also read the post: [Introducing Lazy Evaluation](http://filimanjaro.com/blog/2014/introducing-lazy-evaluation/) thoroughly. 

But there is couple of vague points with respect to adding `_.reduce` to the lazy chain which needs to be discussed a bit.

Let's say we got this:

``` javascript
_([10, 2, 30, 4, 50, 6, 70, 8, 90])
  .map(func1).map(func2).map(func3).value();
```

As it is described in the post, the solution implemented in lodash, when evaluating the code above is like:

``` javascript
var result = [];
for(var i = 0; i < source.length; i++) {
   result[i] = func3(func2(func1(source[i])));
}
```

So far everything is just fine. Now let's say we add a `reduce` call to the chain, like:

``` javascript
_([10, 2, 30, 4, 50, 6, 70, 8, 90])
  .map(func1).map(func2).map(func3)
  .reduce(function(accumulator, n){ return accumulator + '|' + n; })
  .value()//"16|8|36|10|56|12|76|14|96"
```

Considering the above example we could see that the structure of a `reduce` call is different in terms of what it returns (here a `String`) and how it operates, in compared with other lazy operations like `filter`, `map` which return arrays so that we could iterate over them lazily.

This solution (wrapping all the iteratees like `func3(func2(func1(source[i])));`) works for those type of functions in which the type of the return value is usually the same or similar to the input's, like a `map` which usually gets an array and returns another one, so that we could chain several `map`s together.

But when dealing with `reduce` since the type of output is not the same as the input's, we can not call it more than once, I mean it doesn't usually make sense to do:

``` javascript
_(source).reduce(function(){  }).reduce(function(){  }).reduce(function(){  })
```

Unless it return a new object or a new array like:

``` javascript
function iterate(result, value,index){
  result.push('item[' + index + ']:' + value);
  return result;
}
function join(result, str){
  return result + '|' + str;
}

_([10, 2, 30, 4, 50, 6, 70, 8, 90])
  .reduce(iterate, [])//["item[0]:10", "item[1]:2", "item[2]:30"...]
  .reduce(join)//"item[0]:10|item[1]:2|item[2]:30|..."
```

or we change the output ourselves like changing a `string` to an array: 

``` javascript
_([1,2,3])
  .reduce(function(accumulator, val){ return accumulator + '' + val; })//'123'
  .map(identity);//['1','2','3']
  .reduce(function(accumulator, n){ return accumulator + '|' + n; })
  .reduce(function(){  })
```

**Note**: as far as lazy `reduce` is not implemented yet, it won't work with the current version of lodash, but it works in mine since I have implemented it.
# My Approach:

All I want to bring up is, I had to change the algorithm in a way that it could save the old values and also add a new one. And for this example:

``` javascript
_([1,2,3])
  .map(f1).map(f2).map(f3).reduce(f4).map(f5).map(f6).reduce(f7).map(f8).map(f9).value()
```

We now have 2 different type of lazy functions. Those which return arrays like `filter` and `map`, those which doesn't do that, like `reduce`. For now we only have `reduce` in the second list. And considering this point, I split `iteratees` based on the second list and execute them in groups, like:

``` javascript
value = f3(f2(f1(source[i])));

value = reduce(value, f4);

value = f6(f5(value));

value = reduce(value, f7);

value = f9(f8(value));
```

This solution is now working in my own fork of `lodash`, although it is not completed yet.

Describing this issue is just to clarify the differences I have made and also for your **feedback** if there is any, and more importantly if you think there could be **a totally different and better way** of doing it.

**Comment by FilipZawada** [comment]
Isn't `reduce` just an `each`?

PSEUDO-CODE:

```
LazyWrapper.prototype.reduce = function(seed, accumulatorFn) {
     this.each(function(item) {
               seed = accumulatorFn(seed, item)
     })
}
```

I'm not really into the topic now, just popped in out of the blue...

**Comment by mehranhatami** [comment]
@FilipZawada exactly. You just simply made your point. And that's how it defers from other lazy operations like `filter` and `map`.
Considering the fact that it provides a value to the chain which is not usually an **array**, I had to change that specific part of the algorithm which has a **nested iteration** over the `array` and `iteratees` inside the `lazyValue` function:

``` javascript
//line: 1169
outer:
while (length-- && resIndex < takeCount) {// iterates over array
  index += dir;

  var iterIndex = -1,
      value = array[index];

  while (++iterIndex < iterLength) {// iterates over iteratees
    var data = iteratees[iterIndex],
        iteratee = data.iteratee,
        type = data.type;

    //
    var computed = iteratee(value);

    //...
  }
  result[resIndex++] = value;
}
```

Just to get the a better insight here, see how `result` is an array which is used for the whole chain.

**Comment by jdalton** [comment]
Where's the code?

**Comment by mehranhatami** [comment]
@jdalton I just pushed the code. It still has couple of issues, like when the default value of accumulator is not passed and couple of minor issues like that. The main point which needs to be discussed here could be found in `lazyIterate` function.

**Comment by jdalton** [comment]
Thanks. Is there anyway to consolidate this a bit?

**Comment by mehranhatami** [comment]
@jdalton sure there is, and I will continue on that. The code and of course the algorithm is not final!!
In this issue I wanted to first make sure if you see the problem I have, when dealing with non-array2array functions like `reduce` or not.
To consolidate the approach I'd say we need something similar to what is called a `sequence` in [Lazy.js](http://danieltao.com/lazy.js/) of course not exactly the same, just a unified API for **any type of data**, not just arrays.
The point is, lodash already has a `sequence` like API which is called `LazyWrapper`. But as its name suggests it plays its role as a wrapper than a unified API. My path to consolidate the approach is to make some changes on `LazyWrapper` to make it more like an API than just a wrapper. Hope this makes sense to you.

**Comment by jdalton** [comment]
The biggest thing is impact on size here. We want lazy eval to be a nice bonus but not be a burden. lodash isn't a lazy eval lib so it doesn't have to be perfect it just has to hit the high impact areas.

**Comment by mehranhatami** [comment]
@jdalton that's exactly why I use lodash in most of my apps, a perfect match with my concerns. So this is what I do now, I'll try to consolidate the solution and make it stable, without having huge impact on size.

**Comment by jdalton** [comment]
Rock!
